File: go/adapters/http/http.go
package http

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"net/url"

	xrr "hop.top/xrr"
	"gopkg.in/yaml.v3"
)

// Request represents an HTTP interaction request.
type Request struct {
	Method  string            `yaml:"method"  json:"method"`
	URL     string            `yaml:"url"     json:"url"`
	Headers map[string]string `yaml:"headers,omitempty" json:"headers,omitempty"`
	Body    string            `yaml:"body,omitempty"    json:"body,omitempty"`
}

func (r *Request) AdapterID() string { return "http" }

// Response represents an HTTP interaction response.
type Response struct {
	Status  int               `yaml:"status"`
	Headers map[string]string `yaml:"headers,omitempty"`
	Body    string            `yaml:"body,omitempty"`
}

func (r *Response) AdapterID() string { return "http" }

// Adapter implements xrr.Adapter for HTTP interactions.
type Adapter struct{}

func NewAdapter() *Adapter { return &Adapter{} }

func (a *Adapter) ID() string { return "http" }

// Fingerprint: sha256(method + path+query + sha256(body)[:8])[:8].
func (a *Adapter) Fingerprint(req xrr.Request) (string, error) {
	r, ok := req.(*Request)
	if !ok {
		return "", fmt.Errorf("http: unexpected request type %T", req)
	}
	u, err := url.Parse(r.URL)
	if err != nil {
		return "", fmt.Errorf("http: parse url: %w", err)
	}
	pathQuery := u.Path
	if u.RawQuery != "" {
		pathQuery += "?" + u.RawQuery
	}
	bodyHash := sha256.Sum256([]byte(r.Body))
	canonical, err := json.Marshal(map[string]any{
		"method":    r.Method,
		"path":      pathQuery,
		"body_hash": fmt.Sprintf("%x", bodyHash[:4]),
	})
	if err != nil {
		return "", fmt.Errorf("http: fingerprint marshal: %w", err)
	}
	sum := sha256.Sum256(canonical)
	return fmt.Sprintf("%x", sum[:4]), nil
}

func (a *Adapter) Serialize(v any) ([]byte, error)          { return yaml.Marshal(v) }
func (a *Adapter) Deserialize(data []byte, target any) error { return yaml.Unmarshal(data, target) }

File: py/src/xrr/adapters/http.py
"""http adapter — fingerprints on method + path+query + body hash."""
from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass, field
from typing import Any
from urllib.parse import urlparse


@dataclass
class HttpRequest:
    method: str
    url: str
    headers: dict[str, str] = field(default_factory=dict)
    body: str = ""


@dataclass
class HttpResponse:
    status: int
    headers: dict[str, str] = field(default_factory=dict)
    body: str = ""


class HttpAdapter:
    id = "http"

    def fingerprint(self, req: HttpRequest) -> str:
        parsed = urlparse(req.url)
        path_query = parsed.path
        if parsed.query:
            path_query += "?" + parsed.query
        body_hash = hashlib.sha256(req.body.encode()).hexdigest()[:8]
        key = {"method": req.method, "path": path_query, "body_hash": body_hash}
        canonical = json.dumps(key, sort_keys=True, separators=(",", ":"))
        return hashlib.sha256(canonical.encode()).hexdigest()[:8]

    def serialize_req(self, req: HttpRequest) -> dict[str, Any]:
        return {
            "method": req.method,
            "url": req.url,
            "headers": req.headers,
            "body": req.body,
        }

    def serialize_resp(self, resp: HttpResponse) -> dict[str, Any]:
        return {"status": resp.status, "headers": resp.headers, "body": resp.body}

    def deserialize_req(self, data: dict[str, Any]) -> HttpRequest:
        return HttpRequest(
            method=data["method"],
            url=data["url"],
            headers=data.get("headers", {}),
            body=data.get("body", ""),
        )

    def deserialize_resp(self, data: dict[str, Any]) -> HttpResponse:
        return HttpResponse(
            status=data["status"],
            headers=data.get("headers", {}),
            body=data.get("body", ""),
        )
